### DISCUSSION TITLE
writing a modbus client in berry, possible or not?

### DISCUSSION THREAD

**Discussion by ABaumruck** [arendst/Tasmota#19795] [Q&A] [answered]
Hello everybody!
I hope someone can help or point me into the right direction.

I need a sensor at a place where only modbus is available.
Sure I could add something to the sourcecode, but I thought if it is possible to write a "I2C-Driver" in berry, why not a modbus client.

At this moment I do not know if this is even possible.
Your comments are highly welcome.

BR & THX Andreas
BTW I asked this question in the discord channel, but this might be the wrong place

**Comment by sfromis** [comment]
Tasmota already has quite a bit of support for Modbus, are you trying to use that?

I'm not, but to the extent that you get receive messages in the console+MQTT, you could pick such up from Berry, and issue the console commands to send.

**Comment by ABaumruck** [comment]
Thank you for your reply!
Yes, I've digged through the docs and various internet ressources quite intensively.
As far as I understand the existing modbus support of tasmota, it is about modbus "master" or "bridge".
But I couldn't find howto use tasmota as a "modbus client (sensor)".

In my Case I need  the values from a BME280 at a place where only a modbus connection is availabe.
And I'm trying to use tasmota and react to a modbus RTU request, because the software is so mature that I wanted to avoid "reinventing the wheel".

From the first glance I thought berry could be capeable enough to do that (listening to the serial interface, decoding the message and replying) , but in terms of the available ressources, I don't know.

- Do I need to poll when listening to the serial interface
- would the berry code block the rest of the code, and how could I avoid that
- is there a serial interrupt that tells me when a new message has arrived
- etc.

Here I'm stuck. That's why I've asked.
So, do you think my task could be achieved with berry (and tasmota)?
If yes, do you think this is a good idea?
If no, any other suggestions?

 Thx, Andreas

**Reply by sfromis** [reply, marked as answer]
To the extent that you can handle Modbus as "simply serial" at the lower level, you can indeed use Berry to build whatever protocols on top of that. And yes, you need to poll to keep the incoming buffer emptied, like using existing driver callbacks for intervals like 250 ms (or what you choose).

One important thing for doing stuff in Berry is to just not create blocking code, never code loops to wait for something. The Berry code should ideally return in less than 50 ms.

Example of a poller from a driver serving a serial device:
```
  def every_50ms()
    if self.ser.available() <= 0 return end
    var databytes = self.ser.read()
    if databytes.size() == 0 return end
    tasmota.set_timer(0, /->self.process(databytes))
  end
```
This is automatically called when the class instance is registered as a driver. The purpose of the odd-looking `tasmota.set_timer(0, /->` is to get out of the callback as quickly as feasible, passing the data to the called function next iteration of the Tasmota main loop. Using the [bytes](https://github.com/berry-lang/berry/wiki/Chapter-7#bytes-class) object from `serial.read()`, you have a lot of functions for handling structured binary data.

Maybe it could be packaged as a more general-purpose client Modbus driver, but I've not explored details of the feasibility of this, or whether it would work well to extend the existing Modbus support in Tasmota. Arguably, it could make sense to have a Modbus class in Berry, in a similar way as the existing serial class. I'm not using Modbus, but I'd probably look in that direction if I got such an itch to scratch.

**Reply by ABaumruck** [reply]
Thank you!
You have exactely addressed my concerns about using berry for this.
I might give it a try and see how far I can get.
I do not think that my case is a common use case, because my resarch only delivered a handful of pages with "modbus client" code/examples.
BR Andreas

**Reply by FulvioSpelta** [reply]
Hi @ABaumruck did you have done something more structured?
I need too to read some "holding register". Currently i'm reading them using a rule that schedule some "modbusread" commands and get the value from mqtt.

It could be of interest to have in berry these values.
Thanks

Edit:
Sorry found useful info here: #17149 
I'll do some test.
